Dockerfile

```diff
@@ -1,18 +1,12 @@
 # Build Go binary without cgo dependencies
 FROM golang:1.14.3 as builder
 WORKDIR /go/src/github.com/gocardless/theatre
-
-# Clone our fork of envconsul and build it
-RUN set -x \
-      && git clone https://github.com/gocardless/envconsul.git \
-      && cd envconsul \
-      && git checkout 2eb7fdc4dd1a13464e9a529e324ffd9b8d12ce25 \
-      && make linux/amd64 \
-      && mkdir ../bin \
-      && mv pkg/linux_amd64/envconsul ../bin
+COPY go.mod go.sum /go/src/github.com/gocardless/theatre/
+RUN go mod download -x
 
 COPY . /go/src/github.com/gocardless/theatre
-RUN make VERSION=$(cat VERSION) build
+
+RUN make -j 3 VERSION=$(cat VERSION) build
```

Comment:
Made a comment in the other PR- we shouldn't do this, as Go already fires up all the computer has, and sequentially calling it will cache the modules for subsequent builds.

Reply:
Thought I commented in the other one. Increasing the jobs makes a massive difference to me when building all the binaries. I actually call it with more jobs when I am building locally but we thought 3 for the dockerfile was a reasonable compromise. I would be surprised if you didn't also find it faster.

Reply:
You're right about subsequent builds, once your cache is populated. But with a fresh first build and no module cache, starting multiple binary targets building concurrently means we waste resources, and get the job done slower.

This is non-intuitive, especially when you've seen it make things much faster on your local machine. But watch what happens if I build the dockerfile with `time make`, increasing the concurrency (`-j`):

### `-j 1`, 47.8s

```console
Step 6/6 : RUN time make VERSION=$(cat VERSION) build                                                                                                                     
 ---> Running in 73203fc07647                                                                                                                                             
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/rbac-manager ./cmd/rbac-manager/.                                                       
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/workloads-manager ./cmd/workloads-manager/.                                             
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/vault-manager ./cmd/vault-manager/.                                                     
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/theatre-envconsul ./cmd/theatre-envconsul/.                                             
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/theatre-consoles ./cmd/theatre-consoles/.                                               
126.76user 139.77system 0:47.82elapsed 557%CPU (0avgtext+0avgdata 583184maxresident)k                                                                                     
6928inputs+1215272outputs (13major+3039960minor)pagefaults 0swaps  
```

### `-j 2`, 1m 5s

```
Step 6/6 : RUN time make -j 2 VERSION=$(cat VERSION) build
 ---> Running in ec69e68792dc
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/rbac-manager ./cmd/rbac-manager/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/workloads-manager ./cmd/workloads-manager/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/vault-manager ./cmd/vault-manager/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/theatre-envconsul ./cmd/theatre-envconsul/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/theatre-consoles ./cmd/theatre-consoles/.
205.06user 215.02system 1:05.57elapsed 640%CPU (0avgtext+0avgdata 551624maxresident)k 
0inputs+1427232outputs (5major+4390003minor)pagefaults 0swaps
```

### `-j 3`, 1m 37s

```
Step 6/6 : RUN time make -j 3 VERSION=$(cat VERSION) build
 ---> Running in 2b84d4328f0b
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/rbac-manager ./cmd/rbac-manager/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/workloads-manager ./cmd/workloads-manager/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/vault-manager ./cmd/vault-manager/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/theatre-envconsul ./cmd/theatre-envconsul/.
CGO_ENABLED=0 GOARCH=amd64 go build -ldflags "-s -w -X main.Version=1.0.2" -o bin/theatre-consoles ./cmd/theatre-consoles/.
306.67user 331.53system 1:37.39elapsed 655%CPU (0avgtext+0avgdata 582716maxresident)k
61872inputs+1648720outputs (327major+5836035minor)pagefaults 0swaps
```

Hopefully this shows that while it might be great for your dev machine, it won't help our docker setup. Probably the worst bit though, more than this making things generally slower, is that you genuinely do more work as the concurrency increases, as you recompile those vendored modules many times over. So our shared container builder instance will get sad.

Reply:
Ah interesting. Thanks for the detailed response!